File: cmps101/pa5/cs101/List.c
//Ian Feekes
//#1474914
//ifeekes@ucsc.edu
//pa4
//List.c

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include "List.h"
#include <assert.h>
#include <string.h>

Node newNode(int data)
{
 Node N = (Node) malloc(sizeof(struct ListNode));
 N->data = data;
 N->next = NULL;
 N->previous = NULL;
 return(N);
}

List newList(void)
{
 List L = malloc(sizeof(struct ListObject));
 assert(L != NULL);				//when we create a new list it must be NULL
 L->front = NULL;
 L->back = NULL;
 L->cursor = NULL;
 L->length = 0;
 L->index = -1;
 return(L);
}

void freeList(List* pL)
{
 List temp = *pL;
 assert(temp != NULL);  
 if (length(temp) != 0) clear(temp);
 free(*pL);
 *pL = NULL;
}

void freeNode(Node* pN)
{
 free(*pN);
 *pN = NULL;
}

int length(List L)
{
 if(L != NULL) return L->length;
 else
 {
  printf("Error: call to length() returns NULL");
  exit(1);
 }
}

int getIndex(List L)
{
 if(L != NULL)return L->index;
 else
 {
  printf("Error: cannot get the index of a null list");
  exit(1);
 }
}

int front(List L)
{
 if(L != NULL)
 {
  if(length(L) > 0) return(L->front->data);
  else return(-1); 
 }
 else
 {
  printf("Error: cannot get the front a null list");
  exit(1);
 }
}

int back(List L)
{
 if(L != NULL)
 {
  if (length(L) > 0)return(L->back->data);
  else return (-1);
 }
 else
 {
  printf("Error: cannot get the back of a null list");
  exit(1);
 }
}

int getElement(List L)
{
 if(L != NULL)
 {
  if (length(L) > 0 && getIndex(L) >= 0)return(L->cursor->data);
  else return(-1);
 }
 else
 {
  printf("Error: cannot get cursor of a null list");
  exit(1);
 }
}

int equals(List A, List B)
{
 if(A != NULL && B != NULL)
 {
  if(length(A) != length(B))return(0);
  else
  {
   Node tempA = A->front;
   Node tempB = B->front;
   for (; tempA != NULL; tempA = tempA->next)
   {
    if (tempA->data != tempB->data)return(0);
    tempB = tempB->next;
   }
  }
  return(1);
 }
 else
 {
  printf("Error: one of the lists in comparison is null");
  exit(1);
 }
}

void clear(List L)
{
 if(L != NULL)
 {
  for (L->cursor = L->front; L->cursor != NULL;)
  {
   Node temp = L->cursor;
   L->cursor = L->cursor->next;
   free(temp);
  }
  L->front = L->back = L->cursor = NULL;
  L->index = -1;
  L->length = 0;
  }
  else
  {
  printf("Error: list is null");
  exit(1); 
  }
}

void moveTo(List L, int i)
{
 if(L != NULL)
 {
  if (length(L) != 0)
  {
   if (i >= 0 && i <= length(L) -1)
   {
    L->cursor = L->front;
    L->index = 0;
    if (i == 0) return;
    for (; getIndex(L) < i; moveNext(L)){}		//do nothing
   }
  }
 }
 else
 {
  printf("Error: cannot move through an empty list");
  exit(1);
 }
}

void movePrev(List L)
{
 if(L != NULL)
 {
  if (getIndex(L) == 0)
  {
   L->cursor = NULL;
   L->index = -1;
  }
  else if (0 < getIndex(L) && getIndex(L) <= length(L) - 1)
  {
   L->cursor = L->cursor->previous;
   L->index--;
  }
 }
 else
 {
  printf("Error: one of you're lists is improperly declared - movePrev()");
  exit(1);
 }
}

void moveNext(List L)
{
 if(L != NULL)
 {
  if (getIndex(L) == length(L) - 1)
  {
   L->cursor = NULL;
   L->index = -1;
  }
  else if (0 <= getIndex(L) && getIndex(L) < length(L) -1)
  {
   L->cursor = L->cursor->next;
   L->index++;
  }
 }
 else
 {
  printf("Error: one of you're lists is improperly declared - moveNext()");
  exit(1);
 }
}

void prepend(List L, int data)
{
 if(L != NULL)
 {
  if (length(L) == 0)
  {
   Node N = newNode(data);
   L->front = N;
   L->back = N;
   L->length++;
  }
  else if (length(L) > 0)
  {
   Node N = newNode(data);
   N->next = L->front;
   L->front->previous = N;
   L->front = N;
   L->length++;
   if(getIndex(L) != -1)L->index++;
  }
 }
 else
 {
  printf("Error: list is NULl for prepend()");
  exit(1);
 }
}

void append(List L, int data)
{
 if(L != NULL)
 {
  if (length(L) == 0)
  {
   Node N = newNode(data);
   L->front = N;
   L->back = N;
   L->length++;
  }
  else if (length(L) > 0)
  {
   Node N = newNode(data);
   N->previous = L->back;
   L->back->next = N;
   L->back = N;
   L->length++;
  }
 }
 else
 {
  printf("Error: list is NULL for append()");
  exit(1);
 }
}

void insertBefore(List L, int data)
{
 if(L != NULL)
 {
  if(length(L) > 0 && getIndex(L) >= 0)
  {
   Node N = newNode(data);
   Node temp = L->cursor->previous;
   temp->next = N;
   N->previous = N;
   N->next = L->cursor;
   L->cursor->previous = N;
   L->length++;
   L->index++;
  }
 }
 else
 {
  printf("Error: insertBefore called on a null list");
  exit(1);
 }
}

void insertAfter(List L, int data)
{
 if(L != NULL)
 {
  if(length(L) > 0 && getIndex(L) >= 0)
  {
   Node N = newNode(data);
   Node temp = L->cursor->next;
   temp->previous = N;
   N->next = temp;
   N->previous = L->cursor;
   L->cursor->next = N;
   L->length++;
  }	
 }
 else
 {
  printf("Error: insertAfter called on a null list");
  exit(1);
 }
}

void deleteFront(List L)
{
 if(L != NULL)
 {
  if (length(L) >= 2)
  {
   Node temp = L->front->next;
   L->front->next->previous = NULL;
   L->front->next = NULL;
   freeNode(&(L->front));
   L->front = temp;
   L->length--;
   if(getIndex(L) == 0)L->index = -1;
   else if(getIndex(L) != -1) L->index--;
   else if (length(L) == 1)
   {
    Node temp = L->front;
    L->front = NULL;
    L->back = NULL;
    free(temp);
    L->length = 0;
    if(getIndex(L) == 0) L->index = -1;
   }
  }
  else
  {
   printf("Error: deleteFront called on a null list");
   exit(1);
  }
 }
}

void deleteBack(List L)
{
 if(L != NULL)
 {
  if (length(L) >=  2)
  {
   Node temp = L->back->previous; 
   L->back->previous = NULL;
   temp->next = NULL;
   L->back = NULL;
   free(L->back);
   L->back = temp;
   L->length--;
   if(getIndex(L) == length(L) - 1)L->index = -1;
  }
  else if(length(L) == 1)
  {
   Node temp = L->front;
   L->front = NULL;
   L->back = NULL;
   free(temp);
   L->length = 0;
   if(getIndex(L) == 0) L->index = -1;
  }
 }
 else
 {
  printf("Error: deleteBack called on a null list");
  exit(1);
 }
}

void delete(List L)
{
 if(L != NULL)
 {
  if (length(L) > 0 && getIndex(L) >= 0)
  {
   Node temp = L->cursor;
   Node left = L->cursor->previous;
   Node right = L->cursor->next;
   left->next = right;
   right->previous = left;
   L->cursor->next = NULL;
   L->cursor->previous = NULL;
   L->cursor = NULL;
   free(temp);
   L->index = -1;
   L->length--;
  }
 }
 else
 {
  printf("Error: delete called on an empty list");
  exit(1);
 }
}

void printList(FILE* out, List L)
{
 L->cursor = L->front;
 while (L->cursor != NULL)
 {
  fprintf(out,"%i ", (int) (L->cursor->data));
  L->cursor = L->cursor->next;
 }
}

List copyList(List L)
{
 List newL = newList();
 for (Node temp = L->front; temp != NULL; temp = temp->next) append(newL, temp->data);
 return(newL);
}

File: cmps101/pa1/List.java
//Ian Feekes
//#1474914
//ifeekes@ucsc.edu
//PA1
//List.java

public class List
{
 private Node first;
 private Node last;
 private Node cursor; 
 private int length;

 List()//constructor for list. Default is empty
 {
  first = null;
  last = null; 
  cursor = null; 
  length = 0; 
 }

 private class Node 
 {
  int data; 
  Node prev;
  Node next; 

  Node(int d)//constructor for Node
  {
   this.data=d;
   prev=null;//since this is a double linked list we have both a next node and a previous node
   next=null;
  }
 
  public String toString()
  {
   return String.valueOf(data);
  }
 }

 //access functions
 
 int length()//returns number of elements in the list
 {
  return length;
 }

 int index()//if cursor is defined it returns cursor element else it returns -1
 {
  if(cursor!=null)
  { 
   Node A = first;
   int i = 0;
   while(A != null)
   {
    if(A == cursor)return i;
    A = A.next;
    i++;
   }
  }
  return -1;//default is -1 if we can't find anything
 }

 int front()//returns the front element if length() >0
 {
  if(length() > 0) return first.data;
  else throw new RuntimeException("Error: cannot compute front element of an empty list");
 } 

  int back()//returns the back element if length() > 0
  {
   if(length() > 0) return last.data;
   else throw new RuntimeException("Error: cannot compute back element of an empty list");
  }  

 int get()//returns cursor element assuming length and index are greater than or equal to 0
 {
  if((length() > 0)&&(index() >= 0)) return cursor.data;
  else 
  {
   throw new RuntimeException("cannot compute the cursor. Cursor is undefined or list is empty");
  }
 }

 boolean equals(List L)//returns true if the list and L are the same, ignoring the indexes of both
 {
  if(this.length!=L.length) return false;
  Node N = this.first;
  Node O = L.first;
  for(int i = 0; i<this.length; i++)
  {
   if(N.data != O.data) return false;
   N=N.next;
   O=O.next;
  }
  return true;
 }

 //Manipulation Procedures

 void clear()//resets the list to its original empty state
 {
  first = null;
  last = null;
  cursor = null; 
  length = 0;
 }

 void moveFront()//places the cursor to the front of the list if the list isn't empty. If it's empty do nothing
 {
  if(length()==0) return;
  cursor = first;
 }

 void moveBack()//places the cursor to the back of the list if the list isn't empty. If it's empty do nothing
 {
  if(length()==0) return;
  cursor = last;
 }

 void movePrev()//if cursor is defined and not at the front, move it one step closer to the front. If cursor is at the front it becomes undefined. If cursor is undefined do nothing. 
 {
  if(index()==-1) return;
  else if(cursor==first)
  {
   cursor=null;
  } 
  else cursor=cursor.prev;
 }

 void moveNext()//same thing just move it towards the back 
 { 
  if(index()==-1) return;
  else if(cursor==last)
  {
   cursor=null;
  }
  else cursor=cursor.next;
 }

 void prepend(int data)//new element goes at the front of the list. If list is non-empty put it before the front Node
 { 
  Node N = new Node(data);
  if(length()==0)
  {
   first = N;
   last = N;
   length = 1;
  }
  else
  {
  N.next=first;
  first.prev=N;
  first=N;
  length++;
  
  }
 }

 void append(int data)//new element at back
 {
  Node N = new Node(data);
  if(length() == 0)
  {
   first = N;
   last = N;
   length = 1;
  }
  else
  {
   N.prev=last;
   last.next=N;
   last = N;
   length++;
  }
 }

 void insertBefore(int data)//inserts a new element before the cursor assuming the length>=0 && index>=0
 {
  if(cursor==first) prepend(data);
  else if(length()>=0)
  { 
   Node N = new Node(data);
   N.prev = cursor.prev;
   N.prev.next = N; 
   cursor.prev=N;
   length++;
  }
 }

 void insertAfter(int data)//inserts a new element after the cursor assuming the length>=0 && index>=0
 {
  if(cursor==last) append(data);
  else if(length() >0)
  {
   Node N = new Node(data); 
   N.prev = cursor;
   N.next.prev = N;
   cursor.next = N;
   length++;
  }
 }

 void deleteFront()//deletes front element if length>0
 {
  if(length()>0)
  {
   first = first.next;
   first.next.prev=null;
   length--;
  }
   else throw new RuntimeException("Cannot delete the front Node of an empty list");
 }

 void deleteBack()//deletes back element if length>0
 {
  if(length()>0)
  {
   last = last.prev;
   last.prev.next = null;
    length--;
  }
  else throw new RuntimeException("Cannot delete the back Node of an empty list");
 }

 void delete()//deletes cursor element, making the cursor undefined. Pre: length>0, index>=0
 {
  if((length()>0)&&(index()>=0))
  {
   if(cursor==first)
   {
    deleteFront();
    cursor=null;
   }
   else if(cursor==last)
   {
    deleteBack();
    cursor=null;
   }
   else
   {
    cursor.prev.next = cursor.next;
    cursor.next.prev = cursor.prev;
    cursor = null;
    length--;
   }
  }
 }

 //Other Methods:
 
 public String toString()//writes out the list with spaces between each integer(each Node's data)
 {
  Node N = first;
  String str = "";
  while(N!=null)
  {
   str+= (N.data+" ");
   N = N.next;
  }
  return str; 
 }

 List copy()//returns an exact copy of your list, but with an undefined cursor
 {
  List L = new List(); 
  Node N = first;
  while(N!=null)
  {
   L.append(N.data);
   N = N.next;
  }
  return L;
 }

}
